### ISSUE TITLE
Triton installation failed by using pip install source code with pined commit directly

### ISSUE THREAD

**Issue by chuanqi129** [intel/intel-xpu-backend-for-triton#679]
For Pytorch upstream, the triton installation uses pip install source code with pinned commit directly, please refer https://github.com/pytorch/pytorch/blob/main/.ci/docker/common/install_triton.sh#L53. 

I have tried `pip install "git+https://github.com/intel/intel-xpu-backend-for-triton@d67750d5575e6b7bb31b85c20c37f9bc89c720c2#subdirectory=python"` but met below crash issue. **Can we support this installation method also?**

```shell
(pt_39) chuanqiw@mlp-prod-skx-10744:/localdisk/chuanqiw/PT/triton_xpu$ pip install "git+https://github.com/intel/intel-xpu-backend-for-triton@d67750d5575e6b7bb31b85c20c37f9bc89c720c2#subdirectory=python"
Collecting git+https://github.com/intel/intel-xpu-backend-for-triton@d67750d5575e6b7bb31b85c20c37f9bc89c720c2#subdirectory=python
  Cloning https://github.com/intel/intel-xpu-backend-for-triton (to revision d67750d5575e6b7bb31b85c20c37f9bc89c720c2) to /tmp/pip-req-build-abb03q6s
  Running command git clone --filter=blob:none --quiet https://github.com/intel/intel-xpu-backend-for-triton /tmp/pip-req-build-abb03q6s
  Running command git rev-parse -q --verify 'sha^d67750d5575e6b7bb31b85c20c37f9bc89c720c2'
  Running command git fetch -q https://github.com/intel/intel-xpu-backend-for-triton d67750d5575e6b7bb31b85c20c37f9bc89c720c2
  Running command git checkout -q d67750d5575e6b7bb31b85c20c37f9bc89c720c2
  Resolved https://github.com/intel/intel-xpu-backend-for-triton to commit d67750d5575e6b7bb31b85c20c37f9bc89c720c2
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: filelock in /localdisk/chuanqiw/miniconda3/envs/pt_39/lib/python3.9/site-packages (from triton==3.0.0) (3.13.1)
Building wheels for collected packages: triton
  Building wheel for triton (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for triton (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [95 lines of output]
      copy /nfs/site/home/chuanqiw/.triton/nvidia/bin/ptxas to /tmp/pip-req-build-abb03q6s/python/../third_party/nvidia/backend/bin/ptxas ...
      copy /nfs/site/home/chuanqiw/.triton/nvidia/bin/cuobjdump to /tmp/pip-req-build-abb03q6s/python/../third_party/nvidia/backend/bin/cuobjdump ...
      copy /nfs/site/home/chuanqiw/.triton/nvidia/bin/nvdisasm to /tmp/pip-req-build-abb03q6s/python/../third_party/nvidia/backend/bin/nvdisasm ...
      running bdist_wheel
      running build
      running build_py
      running build_ext
      CMake Deprecation Warning at CMakeLists.txt:6 (cmake_policy):
        The OLD behavior for policy CMP0116 will be removed from a future version
        of CMake.
      
        The cmake-policies(7) manual explains that the OLD behaviors of all
        policies are deprecated and that a policy should be set to OLD only under
        specific short-term circumstances.  Projects should be ported to the NEW
        behavior and not rely on setting a policy to OLD.
      
      
      -- The C compiler identification is GNU 9.4.0
      -- The CXX compiler identification is GNU 9.4.0
      -- Detecting C compiler ABI info
      -- Detecting C compiler ABI info - done
      -- Check for working C compiler: /usr/bin/cc - skipped
      -- Detecting C compile features
      -- Detecting C compile features - done
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: /usr/bin/c++ - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      CMake Error at CMakeLists.txt:69 (message):
        Can't find libGenISAIntrinsics.a at
        /nfs/site/home/chuanqiw/.triton/llvm/llvm-4017f04e-ubuntu-x64/lib.  Please
        install it into that directory, or provide explicit path with
        -DGenISAIntrinsics_LIB_DIR=
      
      
      -- Configuring incomplete, errors occurred!
      downloading and extracting https://tritonlang.blob.core.windows.net/llvm-builds/llvm-4017f04e-ubuntu-x64.tar.gz ...
      Traceback (most recent call last):
        File "/localdisk/chuanqiw/miniconda3/envs/pt_39/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/localdisk/chuanqiw/miniconda3/envs/pt_39/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/localdisk/chuanqiw/miniconda3/envs/pt_39/lib/python3.9/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 251, in build_wheel
          return _build_backend().build_wheel(wheel_directory, config_settings,
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 410, in build_wheel
          return self._build_with_temp_dir(
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 395, in _build_with_temp_dir
          self.run_setup()
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 487, in run_setup
          super().run_setup(setup_script=setup_script)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 311, in run_setup
          exec(code, locals())
        File "<string>", line 443, in <module>
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/__init__.py", line 104, in setup
          return distutils.core.setup(**attrs)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 185, in setup
          return run_commands(dist)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
          dist.run_commands()
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
          self.run_command(cmd)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 967, in run_command
          super().run_command(command)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "<string>", line 429, in run
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/wheel/bdist_wheel.py", line 368, in run
          self.run_command("build")
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 967, in run_command
          super().run_command(command)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/command/build.py", line 131, in run
          self.run_command(cmd_name)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 967, in run_command
          super().run_command(command)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "<string>", line 258, in run
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 967, in run_command
          super().run_command(command)
        File "/tmp/pip-build-env-8h79qp94/overlay/lib/python3.9/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "<string>", line 295, in run
        File "<string>", line 372, in build_extension
        File "/localdisk/chuanqiw/miniconda3/envs/pt_39/lib/python3.9/subprocess.py", line 373, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['cmake', '/tmp/pip-req-build-abb03q6s', '-G', 'Ninja', '-DCMAKE_MAKE_PROGRAM=/tmp/pip-build-env-8h79qp94/overlay/bin/ninja', '-DLLVM_SPIRV_DIR=/nfs/site/home/chuanqiw/packages/llvm-spirv', '-DCMAKE_EXPORT_COMPILE_COMMANDS=ON', '-DLLVM_ENABLE_WERROR=ON', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/tmp/pip-req-build-abb03q6s/python/build/lib.linux-x86_64-cpython-39/triton/_C', '-DTRITON_BUILD_TUTORIALS=OFF', '-DTRITON_BUILD_PYTHON_MODULE=ON', '-DPython3_EXECUTABLE:FILEPATH=/localdisk/chuanqiw/miniconda3/envs/pt_39/bin/python', '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON', '-DPYTHON_INCLUDE_DIRS=/localdisk/chuanqiw/miniconda3/envs/pt_39/include/python3.9', '-DTRITON_CODEGEN_BACKENDS=intel', '-DTRITON_PLUGIN_DIRS=', '-DPYBIND11_INCLUDE_DIR=/nfs/site/home/chuanqiw/.triton/pybind11/pybind11-2.11.1/include', '-DLLVM_INCLUDE_DIRS=/nfs/site/home/chuanqiw/.triton/llvm/llvm-4017f04e-ubuntu-x64/include', '-DLLVM_LIBRARY_DIR=/nfs/site/home/chuanqiw/.triton/llvm/llvm-4017f04e-ubuntu-x64/lib', '-DCMAKE_BUILD_TYPE=TritonRelBuildWithAsserts']' returned non-zero exit status 1.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for triton
Failed to build triton
ERROR: Could not build wheels for triton, which is required to install pyproject.toml-based projects
```

**Comment by vlad-penkin** [comment]
@etiotto who can work on this ticket?

**Comment by etiotto** [comment]
So the problem seems to be an installation issue. @vlad-penkin I think @pbchekin team would be the right one to look at this issue.

**Comment by pbchekin** [comment]
It is a kind of blocked by #531: while we are using a custom llvm building with setup.py won't work. Alternatively we can implement a part of `compile-triotn.sh` in `setup.py`, but if we are planning to close #531 shortly, then it is not reasonable to customize `setup.py` more than it is now.

**Comment by etaf** [comment]
@vlad-penkin @riverliuintel  : This issue is blocking the first step in Pytorch Upstream, becase we need to build triton use "pip install" in Pytorch CI,please prioritize, thanks.

**Comment by chuanqi129** [comment]
> It is a kind of blocked by #531: while we are using a custom llvm building with setup.py won't work. Alternatively we can implement a part of `compile-triotn.sh` in `setup.py`, but if we are planning to close #531 shortly, then it is not reasonable to customize `setup.py` more than it is now.

Hi @pbchekin, I saw the #531 has been closed, do we still depends on a custom llvm building? Or can we prepare the custom llvm binaries in advance and download them in setup.py as a shortly solution?

**Comment by pbchekin** [comment]
> Hi @pbchekin, I saw the #531 has been closed, do we still depends on a custom llvm building? Or can we prepare the custom llvm binaries in advance and download them in setup.py as a shortly solution?

We still need a custom llvm, see #717. The existing `python/setup.py` needs `~/packages/llvm-spirv` (https://github.com/intel/intel-xpu-backend-for-triton/blob/llvm-target/python/setup.py#L304) and a custom LLVM in a directory specified by `LLVM_SYSPATH`.

**Comment by pbchekin** [comment]
You can use `./scripts/compile-triton.sh --llvm` to setup `llvm-spirv` and `llvm` directories in advance, then set `LLVM_SYSPATH`. After that `pip install "git+https...` should work. I will verify if that works and will provide additional instructions if needed.

**Comment by chuanqi129** [comment]
> You can use `./scripts/compile-triton.sh --llvm` to setup `llvm-spirv` and `llvm` directories in advance, then set `LLVM_SYSPATH`. After that `pip install "git+https...` should work. I will verify if that works and will provide additional instructions if needed.

Hi @pbchekin, thanks for the reply. That means we need to put the script `./scripts/compile-triton.sh` into pytorch repo firstly, I think it's not an acceptable method for pytorch upstream.

**Comment by pbchekin** [comment]
> That means we need to put the script `./scripts/compile-triton.sh` into pytorch repo firstly, I think it's not an acceptable method for pytorch upstream.

No, i didn't mean that.

Since you mentioned https://github.com/pytorch/pytorch/blob/main/.ci/docker/common/install_triton.sh#L53, then you have an environment where you can set `TRITON_REPO` before running that script. In this environment, you can run `./scripts/compile-triton.sh` that can be downloaded directly from this repository. Anyways, I understand that you do not want to run any custom script from any location, even as a temporary workaround, so we will continue working on `setup.py`.

**Comment by chuanqi129** [comment]
Have verified `pip install "git+https://github.com/intel/intel-xpu-backend-for-triton@111b662bb4e18cc16674d7ca7933a77304ef5a22#subdirectory=python"`, it can install triton 3.0 successfully. Thanks for your great support!

**Comment by leshikus** [comment]
For me fixing `pyproject.toml` helped with `importlib` https://github.com/intel/intel-xpu-backend-for-triton/commit/ff9215ef03c3f028170881969bc1ee60e08554f2

**Comment by pbchekin** [comment]
Fixed in #750.
